store/localcachelayer/user_layer.go

```diff
@@ -66,22 +85,34 @@ func (s LocalCacheUserStore) InvalidateProfilesInChannelCacheByUser(userId strin
 	}
 }
 
-func (s LocalCacheUserStore) InvalidateProfilesInChannelCache(channelId string) {
-	s.rootStore.doInvalidateCacheCluster(s.rootStore.profilesInChannelCache, channelId)
+func (s *LocalCacheUserStore) InvalidateProfilesInChannelCache(channelID string) {
+	s.profilesInChannelMut.Lock()
+	s.profilesInChannelInvalidations[channelID] = true
+	s.profilesInChannelMut.Unlock()
+	s.rootStore.doInvalidateCacheCluster(s.rootStore.profilesInChannelCache, channelID)
 	if s.rootStore.metrics != nil {
 		s.rootStore.metrics.IncrementMemCacheInvalidationCounter("Profiles in Channel - Remove by Channel")
 	}
 }
 
-func (s LocalCacheUserStore) GetAllProfilesInChannel(channelId string, allowFromCache bool) (map[string]*model.User, error) {
+func (s *LocalCacheUserStore) GetAllProfilesInChannel(ctx context.Context, channelId string, allowFromCache bool) (map[string]*model.User, error) {
 	if allowFromCache {
 		var cachedMap map[string]*model.User
 		if err := s.rootStore.doStandardReadCache(s.rootStore.profilesInChannelCache, channelId, &cachedMap); err == nil {
 			return cachedMap, nil
 		}
 	}
 
-	userMap, err := s.UserStore.GetAllProfilesInChannel(channelId, allowFromCache)
+	// If it was invalidated, then we need to query master.
+	s.profilesInChannelMut.Lock()
+	if s.profilesInChannelInvalidations[channelId] {
+		ctx = sqlstore.WithMaster(ctx)
```

Comment:
I guess it's okay and cleaner this way but technically getting the context could be done outside the critical section.

Reply:
Hmm .. how so? Do you mean having an additional flag and then if that flag is true, then set the context? :thinking:

Reply:
Yes, like you are doing with `fromMaster` in some places. I don't know if getting the context is an expensive operation or not so I think it's fine to leave it as is unless we have some data showing we should move it out.

Reply:
Ah I see. This code does not seem to be in the hot path, because it'll only happen for cache misses. And `context.WithValue` only returns a struct, so there's not really that much of a time spent in contention.

```
func WithValue(parent Context, key, val interface{}) Context {
	if parent == nil {
		panic("cannot create context from nil parent")
	}
	if key == nil {
		panic("nil key")
	}
	if !reflectlite.TypeOf(key).Comparable() {
		panic("key is not comparable")
	}
	return &valueCtx{parent, key, val}
}
```

But you are right, I'll make a note to run a load test when this gets merged. Just to confirm.